lib/recursive-methods.rb

```diff
@@ -1,49 +1,75 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n the number passed as arg
+# Space complexity:  O(n) where n the number passed as arg
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError.new() if n < 0
+  return n <= 1 ? 1 : n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where is the number of chars in the string
+# Space complexity:  O(n) where is the number of chars in the string
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return "" if s.empty?
+  return s[-1] << reverse(s[0...-1])
```

Comment:
Technically s[0...-1] creates a new string of length n-1, so this algorithm is actually n^2